File: src/wcc.rs
use std::env;
use std::fs::{self, File};
use std::io::{Write, BufWriter};
use std::path::Path;
use std::process::Command;

const TEMP_RUST_FILE: &str = "/tmp/wcc_temp.rs";

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        eprintln!("Usage: {} <source.wh>", args[0]);
        std::process::exit(1);
    }

    let source_file = &args[1];
    let output_path = Path::new(source_file)
        .file_stem()
        .map(|s| s.to_string_lossy().to_string())
        .unwrap_or_else(|| "a.out".to_string());

    // Run wheec and capture output
    let output = Command::new("wheec")
        .arg(source_file)
        .output()
        .expect("Failed to execute wheec");

    if !output.status.success() {
        eprintln!("Error: wheec failed:\n{}", String::from_utf8_lossy(&output.stderr));
        std::process::exit(1);
    }

    // Write generated Rust code to temp file
    let stdout = String::from_utf8_lossy(&output.stdout);
    let mut lines = stdout.lines();
    lines.next(); // Skip header

    let mut file = BufWriter::new(File::create(TEMP_RUST_FILE).expect("Failed to create temp file"));
    for line in lines {
        writeln!(file, "{}", line).unwrap();
    }
    file.flush().unwrap();

    // Compile it into the final binary
    let status = Command::new("rustc")
        .arg(TEMP_RUST_FILE)
        .arg("-o")
        .arg(&output_path)
        .status()
        .expect("Failed to run rustc");

    cleanup();

    if !status.success() {
        eprintln!("rustc failed to compile the program");
        std::process::exit(1);
    }

    println!("Compiled to: {}", output_path);
}

fn cleanup() {
    if Path::new(TEMP_RUST_FILE).exists() {
        let _ = fs::remove_file(TEMP_RUST_FILE);
    }
}

File: src/wcc.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <cstdio>
#include <cstdlib>
#include <filesystem>
#include <memory>
#include <array>

namespace fs = std::filesystem;

const std::string TEMP_CPP_FILE = "/tmp/wcc_temp.cpp";

int run_command(const std::string& cmd) {
    int ret = std::system(cmd.c_str());
    if (ret == -1) {
        std::cerr << "Failed to run command: " << cmd << "\n";
        return -1;
    }
    return WEXITSTATUS(ret);
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        std::cerr << "Usage: " << argv[0] << " <source.wh>\n";
        return 1;
    }

    std::string source_file = argv[1];

    // Derive output executable name from source filename stem
    std::string output_path = fs::path(source_file).stem().string();
    if (output_path.empty()) output_path = "a.out";

    // Run wheec to generate C++ code
    std::string cmd = "wheec " + source_file;
    std::array<char, 256> buffer;
    std::string wheec_output;

    FILE* pipe = popen(cmd.c_str(), "r");
    if (!pipe) {
        std::cerr << "Failed to execute wheec\n";
        return 1;
    }
    while (fgets(buffer.data(), buffer.size(), pipe)) {
        wheec_output += buffer.data();
    }
    int status = pclose(pipe);
    if (status != 0) {
        std::cerr << "wheec failed with exit code " << status << "\n";
        return 1;
    }

    // Write output to temp cpp file
    std::ofstream ofs(TEMP_CPP_FILE);
    if (!ofs) {
        std::cerr << "Failed to create temporary file\n";
        return 1;
    }
    ofs << wheec_output;
    ofs.close();

    // Compile with g++
    std::string compile_cmd = "g++ " + TEMP_CPP_FILE + " -o " + output_path + " -pthread -std=c++17";
    if (run_command(compile_cmd) != 0) {
        std::cerr << "g++ compilation failed\n";
        fs::remove(TEMP_CPP_FILE);
        return 1;
    }

    // Cleanup temp file
    fs::remove(TEMP_CPP_FILE);

    std::cout << "Compiled to: " << output_path << "\n";
    return 0;
}
